xen: nuke, move to by-name.

SigmaSquadron opened the pull request:
## Description of changes

- Migrate Xen to by-name, since it can't be migrated automatically.
  - This has the side effect of simplifying some Xen stuff.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review pr 345192"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).
  - Intermediate commits will be squashed.

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

CertainLach commented:
Do package sets work with by-name?

SigmaSquadron commented:
> Do package sets work with by-name?

I don't think so. This also sends all `xenPackages` to the top-level attribute. (`xenPackages.xen_4_19` -> `xen_4_19`).

SigmaSquadron commented:
I have reached the conclusion I have no idea what I'm doing.

CertainLach commented on pkgs/top-level/all-packages.nix:
```diff
@@ -34021,10 +34021,18 @@ with pkgs;
-  xen-slim = xenPackages.xen_4_19-slim;
+  buildXenPackage = callPackage ../build-support/xen { };
+
+  inherit (callPackages ../by-name/xe/xen { })
```
You shouldn't directly import by-name packages

inherit (xen) xen_4_19 ...;

Might work, but I don't think we want that.
As I can see, by-name is somehow supposed to work with pkgsets, but I can't figure out how.

SigmaSquadron replied:
That's a straight up copy of the documentation. https://github.com/NixOS/nixpkgs/blob/8637e879e425363b14aa23f3ec1f69e32c9f512f/pkgs/by-name/README.md?plain=1#L183-L201

CertainLach replied:
They don't import from by-name, all by-name packages are automatically discovered and imported

SigmaSquadron replied:
Except if the package.nix refers to multiple versions, in which case, `callPackages` in `all-packages.nix` is required.

CertainLach replied:
I think I understand what this documentation wants us to do, I'll try

CertainLach replied:
Yep, It works for me:

by-name.patch

CertainLach replied:
How did I do that, lol

by-name.patch

CertainLach replied:
Github is acting weird:
https://gist.github.com/CertainLach/5fe65ffe618ee4cab917463b496a39a0

SigmaSquadron replied:
maybe send on the xen room if GH won't let you?

SigmaSquadron replied:
all in favour of just putting `ocamlPackages_4_14` in the generic builder and not looking back? it hasn't changed in years

CertainLach replied:
Not familiar with ocaml packaging in NixOS, can't we just use last version, or some breaking change breaks xen?

SigmaSquadron replied:
the 4.19 build looks fine. if `nixpkgs-review` passes I'm inclined to just send it.

CertainLach replied:
I'll finish my rework of this PR and start nixpkgs-review.

CertainLach replied:
NVM, it is still invalid.
    For each package directory, pkgs.lib.isDerivation pkgs.${name} must be true.

CertainLach replied:
Waiting for nixpkgs-review on https://github.com/CertainLach/nixpkgs/commit/183eae6d051258b85257d45dfde415266b607451

CertainLach commented on pkgs/by-name/xe/xen/default.nix:
default.nix -> package.nix to make it evaluate

SigmaSquadron replied:
Naturally it still doesn't evaluate because the issue is in the generic builder, but at least it might pass one of the failing `vet` checks.

emilazy commented:
Hate to say it, but this is a [violation of the `by-name` contract](https://github.com/NixOS/nixpkgs-vet/blob/1724d597cb5060f3fefa83fbfce57977e832aa32/README.md#validity-checks). cc @infinisil for advice on the correct way to handle it, but the one being used for the Darwin SDK rework that seemed okay was to have the package take e.g. a `xenVersion` argument that the `all-packages.nix` calls override appropriately.

It’s also fine to just not be in `by-name`. It’s not expected that every package can currently fit into the schema.

CertainLach commented:
I don't think contract is breached with buildXenPackage?

emilazy commented:
The problem is that `by-name` packages have to look like `callPackage ../by-name/ab/abcd args` and have to be derivations. With the current `pkgs/by-name/xe/xen` structure, it’s impossible to satisfy both.

With the `xenVersion` thing, you can even get fancy and make e.g. `pkgs/by-name/xe/xen_4_18-slim/package.nix` be `{ xen }: xen.override { xenVersion = "4.18"; xenSlim = true; }` if you really don’t want any `all-packages.nix` entries.

SigmaSquadron commented:
I still don't see what exactly is wrong here?

![xen](https://github.com/user-attachments/assets/84c40931-bfd1-4ee0-8bd5-cea9169e5cdc)

emilazy commented:
But you only satisfy that condition because you violate this one:

> - For each package directory, the `pkgs.${name}` attribute must be defined as `callPackage pkgs/by-name/${shard}/${name}/package.nix args` for some `args`.

SigmaSquadron committed:
xen: drop 4.17

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
xen: turn the generic builder into the buildXenPackage function

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>
Co-authored-by: Yaroslav Bolyukin <iam@lach.pw>

SigmaSquadron committed:
xen: revert minSupportedVersion to the upstream value.

As the builder is generic, more people may be using it, so we should try
to keep this value as close to the upstream source of truth as possible.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
xen: drop update script

It was incredibly slow and will be replaced by something r-ryantm can
run soon.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
.github/labeler.yml: edit the xen-project label to reflect the new paths.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
xen: move to by-name

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>
Co-authored-by: Yaroslav Bolyukin <iam@lach.pw>
Co-authored-by: Emily <vcs@emily.moe>

SigmaSquadron committed:
aliases: explain the removals of xen-slim, the versioned attributes and the xenPackages set

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron commented:
Got an OCaml-related build failure with 4.18.

```console
File "disk.ml", line 33, characters 10-24:
33 |   let c = Char.lowercase c in
               ^^^^^^^^^^^^^^
Error: Unbound value Char.lowercase
make[6]: *** [/build/xen-bd51e57/tools/ocaml/xenstored/../Makefile.rules:31: disk.cmx] Error 2
make[6]: Leaving directory '/build/xen-bd51e57/tools/ocaml/xenstored'
make[5]: *** [/build/xen-bd51e57/tools/ocaml/../../tools/Rules.mk:204: subdir-install-xenstored] Error 2
make[5]: Leaving directory '/build/xen-bd51e57/tools/ocaml'
make[4]: *** [/build/xen-bd51e57/tools/ocaml/../../tools/Rules.mk:199: subdirs-install] Error 2
make[4]: Leaving directory '/build/xen-bd51e57/tools/ocaml'
make[3]: *** [/build/xen-bd51e57/tools/../tools/Rules.mk:204: subdir-install-ocaml] Error 2
make[3]: Leaving directory '/build/xen-bd51e57/tools'
make[2]: *** [/build/xen-bd51e57/tools/../tools/Rules.mk:199: subdirs-install] Error 2
make[2]: Leaving directory '/build/xen-bd51e57/tools'
make[1]: *** [Makefile:64: install] Error 2
make[1]: Leaving directory '/build/xen-bd51e57/tools'
make: *** [Makefile:147: install-tools] Error 2
```

SigmaSquadron commented:
Result of `nixpkgs-review pr 345192` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>20 packages built:</summary>
  <ul>
    <li>xen_4_18</li>
    <li>xen_4_18-slim</li>
    <li>xen_4_18-slim.boot</li>
    <li>xen_4_18-slim.dev</li>
    <li>xen_4_18-slim.doc</li>
    <li>xen_4_18-slim.man</li>
    <li>xen_4_18.boot</li>
    <li>xen_4_18.dev</li>
    <li>xen_4_18.doc</li>
    <li>xen_4_18.man</li>
    <li>xen_4_19</li>
    <li>xen_4_19-slim</li>
    <li>xen_4_19-slim.boot</li>
    <li>xen_4_19-slim.dev</li>
    <li>xen_4_19-slim.doc</li>
    <li>xen_4_19-slim.man</li>
    <li>xen_4_19.boot</li>
    <li>xen_4_19.dev</li>
    <li>xen_4_19.doc</li>
    <li>xen_4_19.man</li>
  </ul>
</details>

CertainLach commented on pkgs/build-support/xen/default.nix:
```diff
@@ -103,11 +107,22 @@ let
 
   # Mark versions older than minSupportedVersion as EOL.
-  minSupportedVersion = "4.17";
+  minSupportedVersion = "4.16";
```
```suggestion
  minSupportedVersion = "4.18";
```

SigmaSquadron replied:
I'd rather keep it at the upstream value. My reasoning is that people maintaining their own downstream Xen derivations might want to use the generic builder to build a 4.16 or 4.17 Xen, and since the updateScript that depended on grepping the value no longer exists, we have no reason to restrict the builder arbitrarily.

CertainLach replied:
Oh, I thought it was a typo during removal of 4.17

CertainLach commented:
> Got an OCaml-related build failure with 4.18. Looks like we'll still have to pin it after all.

We can only pin 4.18 and keep building 4.19 with latest ocaml?

SigmaSquadron commented:
> We can only pin 4.18 and keep building 4.19 with latest ocaml?

Fixed. 4.18 now has an override to pin OCaml to 4.14.

CertainLach commented on pkgs/top-level/all-packages.nix:
```diff
@@ -26550,11 +26550,9 @@ with pkgs;
   });
 
-  # See `xenPackages` source for explanations.
   # Building with `xen` instead of `xen-slim` is possible, but makes no sense.
```
Actually, it allows for much slimmer system closure, as you will no longer have two xen.efi binaries and command line utilities

But I think it is better to just use slim version everywhere by default.

CertainLach reviewed (approved):
Thanks, ofborg!

SigmaSquadron commented:
Result of `nixpkgs-review pr 345192` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>qubes-core-vchan-xen</li>
  </ul>
</details>
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>68 packages built:</summary>
  <ul>
    <li>appvm</li>
    <li>collectd</li>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
    <li>docker-machine-kvm2</li>
    <li>gnome-boxes</li>
    <li>guestfs-tools</li>
    <li>htcondor</li>
    <li>libguestfs</li>
    <li>libguestfs-with-appliance</li>
    <li>librenms</li>
    <li>libvirt</li>
    <li>libvirt-glib</li>
    <li>libvirt-glib.dev</li>
    <li>libvirt-glib.devdoc</li>
    <li>libvmi</li>
    <li>libvmi.dev</li>
    <li>libvmi.lib</li>
    <li>mgmt</li>
    <li>minikube</li>
    <li>multipass</li>
    <li>ocamlPackages.ocaml_libvirt</li>
    <li>perl536Packages.SysVirt</li>
    <li>perl536Packages.SysVirt.devdoc</li>
    <li>perl538Packages.SysVirt</li>
    <li>perl538Packages.SysVirt.devdoc</li>
    <li>python311Packages.guestfs</li>
    <li>python311Packages.guestfs.dist</li>
    <li>python311Packages.libvirt</li>
    <li>python311Packages.libvirt.dist</li>
    <li>python312Packages.guestfs</li>
    <li>python312Packages.guestfs.dist</li>
    <li>python312Packages.libvirt</li>
    <li>python312Packages.libvirt.dist</li>
    <li>qemu_xen</li>
    <li>qemu_xen.debug</li>
    <li>qemu_xen.ga</li>
    <li>rubyPackages.ruby-libvirt</li>
    <li>rubyPackages_3_1.ruby-libvirt</li>
    <li>rubyPackages_3_2.ruby-libvirt</li>
    <li>vagrant</li>
    <li>virt-manager</li>
    <li>virt-manager-qt</li>
    <li>virt-manager.dist</li>
    <li>virt-top</li>
    <li>virt-viewer</li>
    <li>xen (xen_4_19)</li>
    <li>xen-guest-agent</li>
    <li>xen-slim (xen_4_19-slim)</li>
    <li>xen-slim.boot (xen_4_19-slim.boot)</li>
    <li>xen-slim.dev (xen_4_19-slim.dev)</li>
    <li>xen-slim.doc (xen_4_19-slim.doc)</li>
    <li>xen-slim.man (xen_4_19-slim.man)</li>
    <li>xen.boot (xen_4_19.boot)</li>
    <li>xen.dev (xen_4_19.dev)</li>
    <li>xen.doc (xen_4_19.doc)</li>
    <li>xen.man (xen_4_19.man)</li>
    <li>xen_4_18</li>
    <li>xen_4_18-slim</li>
    <li>xen_4_18-slim.boot</li>
    <li>xen_4_18-slim.dev</li>
    <li>xen_4_18-slim.doc</li>
    <li>xen_4_18-slim.man</li>
    <li>xen_4_18.boot</li>
    <li>xen_4_18.dev</li>
    <li>xen_4_18.doc</li>
    <li>xen_4_18.man</li>
  </ul>
</details>

emilazy commented on pkgs/by-name/xe/xen_4_18/package.nix:
```diff
@@ -0,0 +1,48 @@
+
+buildXenPackage.override
+  {
+    inherit slim;
```
Perhaps this should be part of the main parameters rather than something you have to `.override`?

SigmaSquadron replied:
I'm not exactly sure what's the best way to do this, tbh. I need users to be able to override the options in the generic builder, but they can't to `xen.override` because that can only override one input: `xen_4_19`, and they can't do `xen_4_19.override` because that can only override `buildXenPackage`, `ocaml-ng` and `slim`. How do I make sure overrides go all the way to the generic builder, and don't try to override each individual expression?

emilazy replied:
Yeah, I think you’d need to duplicate all that stuff.

My suggestion is to make `pkgs/by-name/xe/xen/package.nix` that contains all the feature flags explicitly and `xenVersion ? "4.19"`, which it uses to import e.g. `./4.18.nix` containing the version‐specific stuff. Then you can do `xen_4_18 = callPackage ../by-name/xe/xen/package.nix { xenVersion = "4.18"; };` in `all-packages.nix`, which is explicitly allowed, and it should work with all the override flags.

This isn’t the most elegant thing in the world, but I think it’s the best way to achieve multi‐version packages like this within the current `by-name` structure that follows the letter of the law and isn’t too hacky.

SigmaSquadron replied:
I really would rather avoid adding anything to `all-packages.nix`. Maybe we should copy all of the feature flags to `4.1{8,9}/package.nix` and that would allow users to override `xen_4_1X`? Not sure about the unpinned `xen` though.

CertainLach replied:
I believe it's ok to use override for isSlim, that's how users expect to override xen, without reading all of those builder implementation details

SigmaSquadron replied:
Yeah, people who need more granularity can write their own Xen derivations.

emilazy replied:
Unfortunately I don’t think we have a good way to do multi‐version packages that doesn’t involve duplication or putting things into `all-packages.nix` at present. `by-name` is designed for the common case and will probably need extension to handle this sort of thing elegantly. But if you really don’t want anything in `all-packages.nix`, even though it’s explicitly okay to use `by-name` stuff like that, then duplication should work.

Note that moving the QEMU stuff into `by-name` poses the same problem for overrides there.

emilazy replied:
I don’t have enough self‐control to stop myself pointing out that dropping the non‐slim Xen would happen to conveniently solve this quandary :)

SigmaSquadron replied:
I really don't think we should remove fat Xen at this time.

emilazy commented on pkgs/by-name/xe/xen_4_19/package.nix:
```diff
@@ -55,4 +41,14 @@ callPackage (import ../generic/default.nix {
-}) ({ ocamlPackages = ocaml-ng.ocamlPackages_4_14; } // genericDefinition)
+  descriptions = {
+    description = "Standard";
+    longDescription = ''
+      Standard version of the Xen Project Hypervisor. Uses forks of QEMU, SeaBIOS,
+      OVMF and iPXE provided by the Xen Project. This provides the vanilla Xen
+      experience, but wastes space and build time. A typical NixOS setup that runs
+      lots of VMs will usually need to build two different versions of QEMU when using
+      this Xen derivation (one fork and upstream).
+    '';
```
Why does this need to be duplicated with the builder?

SigmaSquadron replied:
It... doesn't. I forgot to remove it :P

emilazy commented on pkgs/top-level/aliases.nix:
```diff
@@ -1695,6 +1695,8 @@ mapAliases {
   xcodebuild6 = throw "'xcodebuild6' has been renamed to/replaced by 'xcodebuild'"; # Added 2023-12-10
   xdg_utils = xdg-utils; # Added 2021-02-01
   xen-light = throw "'xen-light' has been renamed to/replaced by 'xen-slim'"; # Added 2024-06-30
+  xenPackages.xen_4_16 = throw "While Xen 4.16 was still security-supported when it was removed from Nixpkgs, it would have reached its End of Life a couple of days after NixOS 24.11 released. To avoid shipping an insecure version of Xen, the Xen Project Hypervisor Maintenance Team decided to delete the derivation entirely."; # Added 2024-09-29
+  xenPackages.xen_4_17 = throw "Due to technical challenges involving building older versions of Xen with newer dependencies, the Xen Project Hypervisor Maintenance Team decided to switch to a 'n+n-1' support cycle. As Xen 4.17 would have been the 'n-2' version, it was removed."; # Added 2024-09-29
```
Hmm, I guess this works as long as `xenPackages` is not meant to exist. But maybe `xenPackages = throw "…use the top‐level packages please…";` is more informative, given people are more likely using the newer versions from this package set if anything anyway?

emilazy commented on pkgs/top-level/all-packages.nix:
```diff
@@ -26550,11 +26550,9 @@ with pkgs;
-  qemu_xen_4_19 = lowPrio (qemu.override { hostCpuTargets = [ "i386-softmmu" ]; xenSupport = true; xen = xenPackages.xen_4_19-slim; });
-  qemu_xen_4_18 = lowPrio (qemu.override { hostCpuTargets = [ "i386-softmmu" ]; xenSupport = true; xen = xenPackages.xen_4_18-slim; });
-  qemu_xen_4_17 = lowPrio (qemu.override { hostCpuTargets = [ "i386-softmmu" ]; xenSupport = true; xen = xenPackages.xen_4_17-slim; });
+  qemu_xen_4_19 = lowPrio (qemu.override { hostCpuTargets = [ "i386-softmmu" ]; xenSupport = true; xen = xen_4_19-slim; });
+  qemu_xen_4_18 = lowPrio (qemu.override { hostCpuTargets = [ "i386-softmmu" ]; xenSupport = true; xen = xen_4_18-slim; });
```
FWIW, these could also go into `by-name` if you wanted (but this is completely non‐blocking and you should use your own judgement).

SigmaSquadron replied:
Moved.

emilazy reviewed (commented):
Broadly looks good; a few things I’m unsure about.

SigmaSquadron committed:
qemu_xen: move to by-name

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron commented:
I'd like to apologise in advance for the things you review here.

SigmaSquadron committed:
nixos/xen: simplify package options

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
treewide: purge remaining calls to xen-slim

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
release-notes: remove mentions to xen-slim or 4.18

They were removed during the great Xen deletion.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
xen: nuke half of the builder

- Removes the non-slim build instructions, massively simplifying
everything in the package.

- Removes unecessary patches.

- Inherits functions from lib instead of repeating lib.* everywhere.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
qemu: have qemu_xen build for i386-softmmu by default

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>
